plugin-maven/src/main/java/com/diffplug/spotless/maven/typescript/Tsfmt.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.maven.typescript;
+
+import java.io.File;
+import java.util.Map;
+
+import org.apache.maven.plugins.annotations.Parameter;
+
+import com.diffplug.spotless.FormatterStep;
+import com.diffplug.spotless.maven.FormatterStepConfig;
+import com.diffplug.spotless.maven.FormatterStepFactory;
+import com.diffplug.spotless.npm.TsConfigFileType;
+import com.diffplug.spotless.npm.TsFmtFormatterStep;
+import com.diffplug.spotless.npm.TypedTsFmtConfigFile;
+
+public class Tsfmt implements FormatterStepFactory {
+
+	@Parameter
+	private String tslintFile;
+
+	@Parameter
+	private String tsconfigFile;
+
+	@Parameter
+	private String vscodeFile;
+
+	@Parameter
+	private String tsfmtFile;
+
+	@Parameter
+	private String typescriptFormatterVersion;
+
+	@Parameter
+	private String typescriptVersion;
+
+	@Parameter
+	private String tslintVersion;
+
+	@Parameter
+	private String npmExecutable;
+
+	@Parameter
+	private Map<String, Object> config;
+
+	@Parameter(defaultValue = "${project.build.directory}", required = true, readonly = true)
```

Comment:
This is what I mean by "buildDir" default.  It appears to not actually be readonly, and it also appears that the default does not work.  One solution is for the "buildDir" to be just a regular nullable parameter, since it seems that's how you're actually using it.

Reply:
This was put in due to the tsftmt step formatter wants to have the build dir. In the Gradle impl that seems to be easier, but here in the Maven this seems to be the way. At least that Parameter(defaultValue = "${project.build.directory}" is used somewhere else (and seems to work with all the tests, except maybe the tsconfig test).

I still think the error message with tsconfig is missleading, cause when I looked at it, it was actually mentioning the found file name of the test.ts, so it does not really seem to be an error with missing inputs.

No time to look at this right now, maybe tomorrow. Thx @nedtwigg

Reply:
Sounds good.  My assertion is that `defaultValue` has no effect, because later on you say "if buildDir is null then set it to some default value".  That logic, of setting a default value if it is null, is fine.  My objection, perhaps mistaken, is that the parameter annotation is claiming to magically set a default value, but it isn't working.  So if it isn't working, it shouldn't be there.

If there are other instances like this in the codebase, then I think they are broken too.

Reply:
Also, I'm skeptical that `new File(".")` will work well as a default value for multiproject setups, but I'm willing to merge and ship this and wait for someone to have a problem and then we can worry about fixing it.

Reply:
I think you will get null if you start this without a maven/gradle setup (e.g. in a unit test straight in IDE). Therefore I set the ".". But might be wrong

Reply:
I tried deleting it, and I saw failures in the maven integration tests, which run using "real maven".